pyresample/bucket/__init__.py

```diff
@@ -0,0 +1,291 @@
+# later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Code for resampling using bucket resampling."""
+
+import dask.array as da
+import xarray as xr
+import numpy as np
+from pyresample._spatial_mp import Proj
+
+
+class BucketResampler(object):
+
+    """Bucket resampling is useful for calculating averages and hit-counts
+    when aggregating data to coarser scale grids.
+
+    Below are examples how to use the resampler.
+
+    Read data using Satpy.  The resampling can also be done (apart from
+    fractions) directly from Satpy, but this demonstrates the direct
+    low-level usage.
+
+    .. code-block:: python
+
+        >>> from pyresample.bucket import BucketResampler
+        >>> from satpy import Scene
+        >>> from satpy.resample import get_area_def
+        >>> fname = "hrpt_noaa19_20170519_1214_42635.l1b"
+        >>> glbl = Scene(filenames=[fname])
+        >>> glbl.load(['4'])
+        >>> data = glbl['4']
+        >>> lons, lats = data.area.get_lonlats()
+        >>> target_area = get_area_def('euro4')
+
+    Initialize the resampler
+
+    .. code-block:: python
+
+        >>> resampler = BucketResampler(adef, lons, lats)
+
+    Calculate the sum of all the data in each grid location:
+
+    .. code-block:: python
+
+        >>> sums = resampler.get_sum(data)
+
+    Calculate how many values were collected at each grid location:
+
+    .. code-block:: python
+
+        >>> counts = resampler.get_count()
+
+    The average can be calculated from the above two results, or directly
+    using the helper method:
+
+    .. code-block:: python
+
+        >>> average = resampler.get_average(data)
+
+    Calculate fractions of occurrences of different values in each grid
+    location.  The data needs to be categorical (in integers), so
+    we'll create some categorical data from the brightness temperature
+    data that were read earlier.  The data are returned in a
+    dictionary with the categories as keys.
+
+    .. code-block:: python
+
+        >>> data = da.where(data > 250, 1, 0)
+        >>> fractions = resampler.get_fractions(data, categories=[0, 1])
+        >>> import matplotlib.pyplot as plt
+        >>> plt.imshow(fractions[0]); plt.show()
+    """
+
+    def __init__(self, target_area, source_lons, source_lats):
+
+        self.target_area = target_area
+        self.source_lons = source_lons
+        self.source_lats = source_lats
+        self.prj = Proj(self.target_area.proj_dict)
+        self.x_idxs = None
+        self.y_idxs = None
+        self.idxs = None
+        self._get_indices()
```

Comment:
This (`_get_indices`) is a lot of logic to happen during init method, but I can see how it would be difficult to change it now.

Reply:
One workaround would be to call `self._get_indices()` in the resampler method calls. `if self.idxs is None: self._get_indices()`.